File: answers/leetcode/Compare Version Numbers.cpp
class Solution {
public:
    int compareVersion(string version1, string version2) {
		size_t prev1 = 0, prev2 = 0;
		while (prev1 < version1.length() && prev2 < version2.length()) {
			size_t i = 0;
			for (i = prev1; i < version1.length(); ++ i) {
				if ('.' == version1[i]) {
					break;
				}
			}
			int v1 = atoi(version1.substr(prev1, i - prev1).c_str());
			prev1 = i + 1;
			i = 0;
			for (i = prev2; i < version2.length(); ++ i) {
				if ('.' == version2[i]) {
					break;
				}
			}
			int v2 = atoi(version2.substr(prev2, i - prev2).c_str());
			prev2 = i + 1;
			if (v1 != v2) {
				//@note precedency
				return (v1 > v2) - (v1 < v2);
			}
		}
		if (version1.size() > prev1) {
			int v = atoi(version1.substr(prev1).c_str());
			//@note
			if (v > 0) {
				return 1;
			}
		}
		if (version2.size() > prev2) {
			int v = atoi(version2.substr(prev2).c_str());
			if (v > 0) {
				return -1;
			}
		}
		return 0;
    }
};

File: answers/leetcode/Compare Version Numbers.py
class Solution(object):
    def compareVersion(self, version1, version2):
        """
        :type version1: str
        :type version2: str
        :rtype: int
        """

        version1 = version1.split(".")
        version2 = version2.split(".")
        length1 = len(version1)
        length2 = len(version2)
        for i in range(max(length1, length2)):
            num1 = None
            if i < length1:
                num1 = int(version1[i])
            num2 = None
            if i < len(version2):
                num2 = int(version2[i])
            if (i < length1 and i < length2 and num1 > num2) or (i >= length2 and num1 > 0):
                return 1
            elif (i < length1 and i < length2 and num1 < num2) or (i >= length1 and num2 > 0):
                return -1
        return 0
